### PR TITLE
Refactor REPL result printing

### PR DESCRIPTION
**Pull request by drcicero**
Hi! This is my first pull request on github :)

I changed the handling printing information. Previously, evaluation would either return a single NamedVal or a list of NamedVals, and those could then be either wrapped in a ResultToPrint, and stored in the result_to_print buffer, together with additional LazyPrints elements. Further, ResultToPrint does contain a boolean to tell whether to actually print or not, and there is also a function that filters the list to not print NamedVals of same name twice.

Now evaluating a tree always returns a list of ResultToPrint, which is added to the result_to_print buffer. To ensure all elements of result_to_print are ResultToPrint, LazyPrints are also wrapped into a ResultToPrint. After each command printings are either SILENTly ignored, IMMEDIATEly printed, or BATCHed until the end of evaluating a file. Filtering duplicated NamedVals is handled by setting the respective boolean flag per ResultToPrint (whether to print or not) to false.

Let me know what you think. If its not going in the direction you intended, that's also fine :)

### CODE REVIEW

**Comment by drcicero** [comment]
not sure. which line?

File: RASP_support/REPL.py [hunk 1]
```diff
@@ -219,31 +216,34 @@ def loadFile(self,ast,calling_env=None):
...
-	def run_given_line(self,line):
+	def run_tree(self,fromfile,fromline,env):
 		try:
-			tree = LineReader(given_line=line).get_input_tree()
+			tree = LineReader(fromfile=fromfile,fromline=fromline).get_input_tree()
 			if isinstance(tree,Stop):
-				return None
-			rp = self.evaluate_tree(tree)
-			if isinstance(rp,LazyPrint):
-				rp.print() # error messages get raised, but ultimately have to be printed somewhere if not caught? idk
```
Comment (drcicero) [inline RASP_support/REPL.py:242]: do you mean this line?

Comment (gailweiss) [reply RASP_support/REPL.py:242]: no, what is now line 232 (and was lines 244-245). Basically, the new function run_tree has one way of dealing with antlr exceptions (the 'normal' way for while the REPL is running), but it is replacing the function run_given_line which was originally just for REPL initiation, and which had an extra print if it got an antlr exception (noting which initiating line failed). I don't really mind, I was just wondering if it was a deliberate choice. 

independently, how did you put the diff inside the comment?

Comment (drcicero) [reply RASP_support/REPL.py:242]: Oh, I guess I didn't get the difference between initialization ANTLR errors and code ANTRL error messages, the code is new to me. I am also happy with anything that you do with this :)

For code comments, go to the top of this github pull request page, switch tabs from "conversation" to "commit" at the top, hover over line, press [+], add comment.
i think, you can also do full "code reviews" in the sense that you go line by line through the changes and comment everything you find comment-worthy.

Comment (drcicero) [reply RASP_support/REPL.py:242]: and eventually mark it as resolved ↓

**Comment by gailweiss** [comment]
Hi, I really appreciate the refactorisation! Unfortunately something seems to be wrong: it is dropping prints. For example, now, if I enter a non-existing variable `"dfgdgdg;"` into the REPL (which should spit out 		
```
!!ignoring input:
	 Error: Undefined variable: dgdgdgf" ) 
```
),

it now simply ignores it silently. Similarly for example if I write `1("hi")` (which should print `not implemented yet`). I'm not sure why this is

File: RASP_support/REPL.py [hunk 2]
```diff
@@ -496,11 +461,11 @@ def nice_matrix_line(m):
...
-	return a,None
\ No newline at end of file
+#def runner():
+#	a = REPL()
+#	try:
+#		a.run()
+#	except Exception as e:
+#		print(e)
+#		return a,e
+#	return a,None
```
Comment (gailweiss) [inline RASP_support/REPL.py]: I'd rather leave this uncommented (this is how I run the REPL in python, so I can poke around the environment and errors if something is being weird), can rename it to debug_runner if want to be explicit though

**Review by gailweiss — CHANGES_REQUESTED** [review]
Hi, I really appreciate the refactorisation! Unfortunately something seems to be wrong: it is dropping prints. For example, now, if I enter a non-existent variable "dfgdgdg;" into the REPL (which should spit out

!!ignoring input:
	 Error: Undefined variable: dfgdgdg 

),

it now simply ignores it silently. Similarly for example if I write 1("hi") (which should print not implemented yet). I'm not sure why this is

File: RASP_support/REPL.py (hunk 2 above)
Comment (drcicero) [reply RASP_support/REPL.py]: makes sense

**Comment by drcicero** [comment]
two more commits and i think, errors are visible again! that was an unfortunate combination of passing the wrong number of arguments to ResultToPrint and an "catch all Exception" handler, which in turn contained the same error, or something

**Comment by drcicero** [comment]
ah, not quite, there can also be `None`s in the result_to_print list, hm...

**Review by gailweiss — CHANGES_REQUESTED** [review]
Right now, Silent mode (which I think can be seen as import or loop mode) is supressing errors, it should only be suppressing outputs.

For example, if you add the line `'dfdgd;'` to `rasplib.rasp`, in the 'main' repo it will loudly complain and tell you that it failed to load the file. But if you add it in this repo, it will silently ignore the line.

**Comment by drcicero** [comment]
ah, showing errors always makes sense :)
should i, or did you already start doing something?

**Comment by gailweiss** [comment]
> ah, showing errors always makes sense :)
> should i, or did you already start doing something?

please do :) I am working on making a little test suite!

**Comment by drcicero** [comment]
Tests are great! I made errors to be printed always, and squashed all those stupid small commits into one. What do you think?

**Review by gailweiss — APPROVED** [review]

**Review by gailweiss — CHANGES_REQUESTED** [review]
Hi, thanks for fixing the first change! I'm requesting a couple more changes, I've added some tests that cover them (and then also a bunch of other tests while I was at it---I'm slowly trying to cover everything). 

They are:
1. I'd rather not print an additional newline after a user enters "exit" (or similar) in the interpreter, so I do differentiate this from EOL
2. When loading a `.rasp` file, I print the (filtered) assignments in the file only after the full file is finished loading. This way if I set a variable multiple times in a file, only the last assignment to it is printed. (By the way, I have not implemented this vision perfectly: some of the errors seem to rise up and get printed before everything else, even though I would like them to be printed in order with all the other prints created by the file. Independently of the scope of this PR, I would be glad for a clean-up that fixes this! I've placed a note in `tests/broken_libs/lib/other_errors.rasp` if you are curious. Otherwise, the tests avoid this for now, so as not to enforce this mistake.)



On a personal note: I apologise if all these changes feel nitpicky. I'm trying to strike a balance between making it easy to contribute, and also keeping the code in a way that I can maintain it going forward. Unfortunately by being my first contributor you are hitting all the decisions I didn't even realise I had made! I am now trying to build better tests so you don't have to get so many change requests in the future... sorry about the mess you're dealing with this time

**Comment by drcicero** [comment]
Aha, I suppose that explains this TODO about not all errors having been propagated! It also explains why the thing i called BATCHed printing exists, while I didn't see where its used :)

**Comment by gailweiss** [comment]
hey there! i don't think you're coming back to these changes just yet, so i'm closing this for now :)

**Pull request closed without merging**